### ISSUE TITLE
respect and accept license files during build

### ISSUE THREAD

**Issue by marcoceppi** [juju/charm-tools#136]
Layer and Interface authors should include COPYRIGHT files in each of their respective repositories either using DEP5[0] format or another means (just the copyright license text). charm-build should then be able to parse the copyright file if it is DEP5[0] or determine license type from license text and generate a COPYRIGHT file for the built charm using the DEP5[0] format to properly attribute and respect licensing of components. Intended goal is to have all layers include a COPYRIGHT file and generate the following example from that file:

##layer-foo example

```
├── config.yaml
├── COPYRIGHT
├── layer.yaml
├── reactive
│   └── foo.py
└── wheelhouse.txt
```

### layer-foo/COPYRIGHT

```
Format: http://dep.debian.net/deps/dep5/

Files: *
Copyright: Copyright 2016, Marco Ceppi, All Rights Reserved.
License: MIT
  MIT LICENSE TEXT
```

## layer-charm
```
├── config.yaml
├── COPYRIGHT
├── layer.yaml
├── UPSTREAM_DIST.tar.gz
└── reactive
    └── charm.py

```

### layer-charm/COPYRIGHT

```
Format: http://dep.debian.net/deps/dep5/

Files: *
Copyright: Copyright 2016, Guy LeDouce, All Rights Reserved.
License: LGPL

Files: UPSTREAM_DIST.tar.gz
Copyright: Copyright 2014, Upstream Project, All Rights Reserved.
License: Special Company License
```

## interface-noop
```
├── COPYRIGHT
├── provides.py
├── requires.py
└── peers.py
```

### interface-noop/COPYRIGHT
```
APACHE LICENSE TEXT
```

All of these, once built, should produce

## charm
```
├── config.yaml
├── COPYRIGHT
├── hooks
│   ├── install
│   ├── interfaces
│   │   └── noop
│   │         ├── provides.py
│   │         ├── provides.py
│   │         └── peers.py
│   └── upgrade-charm
├── layer.yaml
├── UPSTREAM_DIST.tar.gz
├── reactive
│   ├── charm.py
│   └── foo.py
└── wheelhouse.txt
```

### charm/COPYRIGHT
```
Format: http://dep.debian.net/deps/dep5/

Files: hooks/* Makefile config.yaml layer.yaml .build.manifest wheelhouse.txt
Copyright: Copyright 2016, Canonical, Ltd., All Rights Reserved.
License: <BASE LAYER LICENSE>

Files: reactive/foo.py
Copyright: Copyright 2016, Marco Ceppi, All Rights Reserved.
License: MIT

Files: reactive/charm.py
Copyright: Copyright 2016, Guy LeDouce, All Rights Reserved.
License: LGPL

Files: hooks/interfaces/noop/*
Copyright: Copyright 2016, INTERFACE AUTHOR
License: Apache

Files: UPSTREAM_DIST.tar.gz
Copyright: Copyright 2014, Upstream Project, All Rights Reserved.
License: Special Company License
```

[0]: http://dep.debian.net/deps/dep5/

**Comment by bcsaller** [comment]
I agree that GPL.x might cause issues for some consumers, but we should insist on a permissive license going forward, not add tooling around a problem that only gets more complex as you explore the space. What if two licenses are not compatible? Can we accept GPL code into an apache project (no) but we can accept apache code into a GPL project, do we detect that? What if some files are one license but overridden in another layer, is that a derived work? Lets just ask that contributors honor the Apache license and we include in base and other important layers a note saying that merges to this repo require copyright assignment or something like that.

**Comment by johnsca** [comment]
@bcsaller We're not expecting built charms to be part of any project, only the source layers, though.

**Comment by bcsaller** [comment]
I was just saying that this gets complex and silly quickly and we should avoid the issue with a simpler policy rather than more tooling (which ultimately won't address all the possible issues)

**Comment by chuckbutler** [comment]
requiring copyright means requiring a CLA right? 

And this CLA would really only apply to the base layer, as anything deriving from that is under a potentially different copyright, and at that point the licensing soup may continue - as thats up to the intent of the author of the derived work?

Is this something that plugging https://www.clahub.com/ into layer:basic repo, giving it a permissive license, and calling this "done for now" would work?

**Comment by Vultaire** [comment]
I'd like to bring this up, or at least, one particular case.

What if a charm was intended to be GPLv3, but because it didn't specify both a LICENSE and copyright file, a built charm ended up claiming itself to be Apache 2 licensed instead?

At present, layer:basic will introduce a LICENSE and copyright file (both Apache2, the latter claiming to apply to all files).  As "charm build" does not warn about missing LICENSE or copyright files, charms may unintentionally be built declaring an Apache 2 license.

Should we perhaps detect the lack of a license and, if not specified, refuse to build?
